Core/Scripting: implement Conversation scripts

funjoker opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

-  Option to make ConversationScripts
-  Added ScriptName column to conversation_template

**Target branch(es):** 3.3.5/master

- [x] master

**Issues addressed:** Closes #  (insert issue tracker number)


**Tests performed:** (Does it build, tested in-game, etc.)
Same tests as in #20757 

**Known issues and TODO list:** (add/remove lines as needed)


<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

funjoker committed:
Add ConversationScripts

Traesh commented on src/server/game/Scripting/ScriptMgr.h:
```diff
@@ -879,6 +880,17 @@ class TC_GAME_API SceneScript : public ScriptObject
+
+public:
+
+    virtual void OnConversationCreate(Conversation* /*conversation*/, Unit* /*creator*/) { }
```
Nice PR, but you should not rely on user calling "SetValidState", instead i think you should make "OnConversationCreate" return a boolean (default true) and set Conversation state using this boolean, so the user just have to return true or false from his script

joschiwald commented on src/server/game/Entities/Conversation/Conversation.cpp:
```diff
@@ -138,6 +140,16 @@ bool Conversation::Create(ObjectGuid::LowType lowGuid, uint32 conversationEntry,
+        return false;
+
+    // option to prevent AddToMap by external script, e.g. ConversationScript or InstanceScript
+    if (!_isValid)
```
like i already said to you, don't use such boolean that the user have to set
the crash does not happen cause of missing actor on index 0, it happens because of the line does not have an actor
you need to verifiy that your conversation have actors for all lines, if that fails, add a log message

funjoker replied:
Since the data is taken by sniff and it is unkown how many actors are there until the developer adds actors, i donno how to Check it if you dont know it.

joschiwald replied:
you know it, actors depend on lines

funjoker replied:
Ah yeah ok. Gonna add a Check for it.

joschiwald commented on src/server/game/Scripting/ScriptMgr.cpp:
```diff
@@ -113,6 +115,10 @@ template<>
+template<>
+struct is_script_database_bound<ConversationScript>
+: std::true_type { };
+
```
did you check that hotswap support too?

funjoker replied:
Never worked with hotswap will read it

funjoker committed:
More general checks
- Remove validState
- Add general valid check
- Hotswap code (correct?)
- Make hook to boolean

funjoker commented on src/server/game/Scripting/ScriptMgr.cpp:
```diff
@@ -778,6 +784,35 @@ class ScriptRegistrySwapHooks<SpellScriptLoader, Base>
     bool swapped;
 };
 
+/// This hook is responsible for swapping ConversationScript's
```
Never worked with hotswap.
If it's incorrect feel free to adjust :)

funjoker committed:
Add some more hooks

funjoker committed:
Copy & Paste issues

joschiwald commented on src/server/game/Entities/Conversation/Conversation.cpp:
```diff
@@ -135,9 +137,25 @@ bool Conversation::Create(ObjectGuid::LowType lowGuid, uint32 conversationEntry,
+    }
+
+    _totalActorIdx.sort();
+    _totalActorIdx.unique();
```
use std::set to prevent duplicates and they are sorted by default

funjoker committed:
std::list -> std::set

Shauren commented on src/server/game/Entities/Conversation/Conversation.cpp:
```diff
@@ -137,18 +137,24 @@ bool Conversation::Create(ObjectGuid::LowType lowGuid, uint32 conversationEntry,
     }
 
     uint16 linesIndex = 0;
+    std::list<uint16> _totalActorIdx;
```
local var naming

funjoker replied:
si senor

funjoker committed:
local naming

funjoker committed:
Putting it into DB PR

joschiwald commented on src/server/game/Scripting/ScriptMgr.h:
```diff
@@ -879,6 +880,24 @@ class TC_GAME_API SceneScript : public ScriptObject
+        virtual void OnAddActor(Conversation* /*conversation*/, ObjectGuid /*actorGuid*/, uint16 /*actorIdx*/) { }
+
+        // Called when a Participant is added to conversation
+        virtual void OnAddParticipant(Conversation* /*conversation*/, ObjectGuid /*participantGuid*/) { }
```
can use provide an example where OnAddActor and OnAddParticipant is useful for?

funjoker replied:
No idea yet. But thought it might be useful for someone

joschiwald commented on src/server/game/Globals/ConversationDataStore.cpp:
```diff
@@ -137,6 +138,7 @@ void ConversationDataStore::LoadConversationTemplates()
             conversationTemplate.Id                 = fields[0].GetUInt32();
             conversationTemplate.FirstLineId        = fields[1].GetUInt32();
             conversationTemplate.LastLineEndTime    = fields[2].GetUInt32();
+            conversationTemplate.ScriptId           = sObjectMgr->GetScriptId(fields[3].GetCString());
```
GetString()

funjoker committed:
GetCString-> GetString

funjoker committed:
Remove useless code

funjoker committed:
Merge remote-tracking branch 'remotes/upstream/master' into conversationScripts

funjoker committed:
Move Hook Call to correct spot

funjoker committed:
Oops

funjoker committed:
Clean Up

funjoker committed:
Another cleaning

joschiwald committed:
prepare merge

Shauren commented on src/server/scripts/World/conversation_scripts.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2008-2017 TrinityCore <http://www.trinitycore.org/>
+ * Copyright (C) 2008-2017 TrinityCore <https://www.trinitycore.org/>
```
lul, im pretty sure its http in all headers

joschiwald replied:
should be changed on next copyright update

joschiwald committed:
fix

joschiwald commented:
merging when travis succeeds
